package NYT.NControllerAgent.NProto;

import "yt/ytlib/scheduler/proto/job.proto";
import "yt/core/misc/proto/guid.proto";

////////////////////////////////////////////////////////////////////////////////

message TJobMetrics
{
    message TEntry
    {
        required int64 key = 1;
        required int64 value = 2;
    }
    message TCustomEntry
    {
        required string statistics_path = 1;
        required string profiling_name = 2;
        required int64 value = 3;
    }

    repeated TEntry values = 1;
    repeated TCustomEntry custom_values = 2;
}

message TTreeTaggedJobMetrics
{
    required string tree_id = 1;
    optional TJobMetrics metrics = 2;
}

////////////////////////////////////////////////////////////////////////////////

message TReqGetOperationInfo
{
    required NYT.NProto.TGuid operation_id = 1;
}

message TRspGetOperationInfo
{
    required bytes progress = 1;       // YSON
    required bytes brief_progress = 2; // YSON
    required bytes running_jobs = 3;   // YSON
    required bytes job_splitter = 4;   // YSON
    required int64 controller_memory_usage = 5;
    required int32 controller_state = 6;
    required bytes alerts = 7; // YSON
}

////////////////////////////////////////////////////////////////////////////////

message TReqGetJobInfo
{
    required NYT.NProto.TGuid operation_id = 1;
    required NYT.NProto.TGuid job_id = 2;
}

message TRspGetJobInfo
{
    required bytes info = 1; // YSON
}

////////////////////////////////////////////////////////////////////////////////

message TPoolTreeControllerSettings
{
    required string tree_name = 1;
    required string scheduling_tag_filter = 2;
    required bool tentative = 3;
}

message TPoolTreeControllerSettingsMap
{
    repeated TPoolTreeControllerSettings tree_settings = 1;
}

message TOperationDescriptor
{
    required NYT.NProto.TGuid operation_id = 1;
    required int32 operation_type = 2; // EOperationType
    required bytes spec = 3; // YSON
    required uint64 start_time = 4; // TInstant
    required string authenticated_user = 5;
    optional bytes secure_vault = 6; // YSON
    reserved 7; // deprecated
    required NYT.NProto.TGuid user_transaction_id = 8;
    required TPoolTreeControllerSettingsMap pool_tree_controller_settings_map = 9;
    required bytes acl = 10; // YSON
}

message TReqRegisterOperation
{
    required NYT.NProto.TGuid incarnation_id = 1;
    required TOperationDescriptor operation_descriptor = 2;
}

message TRspRegisterOperation
{ }

////////////////////////////////////////////////////////////////////////////////

message TControllerTransactionIds
{
    optional NYT.NProto.TGuid async_id = 1;
    optional NYT.NProto.TGuid input_id = 2;
    optional NYT.NProto.TGuid output_id = 3;
    optional NYT.NProto.TGuid debug_id = 4;
    optional NYT.NProto.TGuid output_completion_id = 5;
    optional NYT.NProto.TGuid debug_completion_id = 6;
    repeated NYT.NProto.TGuid nested_input_ids = 7;
}

message TInitializeOperationResult
{
    required bytes mutable_attributes = 1; // YSON
    required bytes brief_spec = 2; // YSON
    required bytes unrecognized_spec = 3; // YSON
    required bytes full_spec = 4; // YSON
    required TControllerTransactionIds transaction_ids = 5;
}

message TReqInitializeOperation
{
    required NYT.NProto.TGuid incarnation_id = 1;
    required NYT.NProto.TGuid operation_id = 2;
    required bool clean = 3;
    optional TControllerTransactionIds transaction_ids = 4;
}

message TRspInitializeOperation
{
    // If the result is missing, it will be sent later in a heartbeat.
    optional TInitializeOperationResult result = 1;
}

////////////////////////////////////////////////////////////////////////////////

message TPrepareOperationResult
{
    optional bytes attributes = 1; // YSON
}

message TReqPrepareOperation
{
    required NYT.NProto.TGuid incarnation_id = 1;
    required NYT.NProto.TGuid operation_id = 2;
}

message TRspPrepareOperation
{
    // If the result is missing, it will be sent later in a heartbeat.
    optional TPrepareOperationResult result = 1;
}

////////////////////////////////////////////////////////////////////////////////

message TMaterializeOperationResult
{
    // Set to true if suspend_operation_after_materialization is set.
    required bool suspend = 1;
    required NScheduler.NProto.TJobResources initial_needed_resources = 2;
}

message TReqMaterializeOperation
{
    required NYT.NProto.TGuid incarnation_id = 1;
    required NYT.NProto.TGuid operation_id = 2;
}

message TRspMaterializeOperation
{
    // If the result is missing, it will be sent later in a heartbeat.
    optional TMaterializeOperationResult result = 1;
}

////////////////////////////////////////////////////////////////////////////////

message TReviveOperationResult
{
    message TRevivedJob
    {
        required NYT.NProto.TGuid job_id = 1;
        required int32 job_type = 2; // EJobType
        required uint64 start_time = 3; // TInstant
        required NScheduler.NProto.TJobResources resource_limits = 4;
        required bool interruptible = 5;
        required string tree_id = 6;
        required uint32 node_id = 7;
        required string node_address = 8;
    }

    required bytes attributes = 1; // YSON
    required bool revived_from_snapshot = 2;
    repeated TRevivedJob revived_jobs = 3;
    repeated string revived_banned_tree_ids = 4;
    required NScheduler.NProto.TJobResources needed_resources = 5;
}

message TReqReviveOperation
{
    required NYT.NProto.TGuid incarnation_id = 1;
    required NYT.NProto.TGuid operation_id = 2;
}

message TRspReviveOperation
{
    // If the result is missing, it will be sent later in a heartbeat.
    optional TReviveOperationResult result = 1;
}

////////////////////////////////////////////////////////////////////////////////

message TCommitOperationResult
{ }

message TReqCommitOperation
{
    required NYT.NProto.TGuid incarnation_id = 1;
    required NYT.NProto.TGuid operation_id = 2;
}

message TRspCommitOperation
{
    // If the result is missing, it will be sent later in a heartbeat.
    optional TCommitOperationResult result = 1;
}

////////////////////////////////////////////////////////////////////////////////

message TReqCompleteOperation
{
    required NYT.NProto.TGuid incarnation_id = 1;
    required NYT.NProto.TGuid operation_id = 2;
}

message TRspCompleteOperation
{ }

////////////////////////////////////////////////////////////////////////////////

message TReqTerminateOperation
{
    required NYT.NProto.TGuid incarnation_id = 1;
    required NYT.NProto.TGuid operation_id = 2;
    required int32 controller_final_state = 3; // EControllerState
}

message TRspTerminateOperation
{ }

////////////////////////////////////////////////////////////////////////////////

message TReqWriteOperationControllerCoreDump
{
    required NYT.NProto.TGuid operation_id = 1;
}

message TRspWriteOperationControllerCoreDump
{
    required string path = 1;
}

////////////////////////////////////////////////////////////////////////////////

message TReqUnregisterOperation
{
    required NYT.NProto.TGuid incarnation_id = 1;
    required NYT.NProto.TGuid operation_id = 2;
}

message TRspUnregisterOperation
{
    repeated TTreeTaggedJobMetrics residual_job_metrics = 1;
}

////////////////////////////////////////////////////////////////////////////////

message TReqUpdateOperationRuntimeParameters
{
    required NYT.NProto.TGuid incarnation_id = 1;
    required NYT.NProto.TGuid operation_id = 2;

    required bytes parameters = 3;
}

message TRspUpdateOperationRuntimeParameters
{ }

////////////////////////////////////////////////////////////////////////////////

